// @hyperledger/firefly-sdk/dist/interfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FireFlyDataBlobRequestDefaults = exports.FireFlyIdempotencyError = exports.FireFlyError = void 0;
/**
 * Most types in this file are aliased from request/response body types that
 * are generated from the OpenAPI specification for FireFly.
 *
 * Because the spec doesn't accurately reflect "required" fields, currently
 * all request types below mark every field as optional, and response types
 * mark every field as being set. This is not completely accurate, but should
 * be close enough for most use cases.
 */
// General
class FireFlyError extends Error {
    constructor(message, originalError, path) {
        super(message);
        this.originalError = originalError;
        this.path = path;
    }
}
exports.FireFlyError = FireFlyError;
class FireFlyIdempotencyError extends FireFlyError {
}
exports.FireFlyIdempotencyError = FireFlyIdempotencyError;
exports.FireFlyDataBlobRequestDefaults = {
    autometa: 'true',
};
const balances = [];
const approvals = [];
